### ISSUE TITLE
Doesn't allow dicts with keys that contain separator in them

### ISSUE THREAD

**Issue by msabramo** [akesterson/dpath-python#30]
I'm using the default separator of `/`. I can't retrieve anything if any keys have a `/` in them. I could change the separator but might run into another problem. I suspect that it could be modified to not care about what data is in the keys.

```
  File "/Users/marca/dev/surveymonkey/InventorySvc/.tox/py27/lib/python2.7/site-packages/dpath/path.py", line 57, in validate
    separator))
InvalidKeyName: /dev/sdc6 at ansible_facts/ohai_filesystem contains the separator /
```

**Comment by akesterson** [comment]
The problem with ignoring separators in keys is that we can no longer properly fetch items in a path. Consider this dict:

    x = {
        "a": {
                "b/c": {
                    "d": 0
                }
            }
    }
    dpath.util.get("/a/b/c/d")

This simple scenario presents an immense problem. In order to properly search into the dictionary, we must tokenize the path given to us. This will allow us to properly glob for keys at any given level of the dictionary. So we tokenize out as ['a', 'b', 'c', 'd']. In this case, we are asking dpath to find us the object 4 levels deep in the dictionary following a strict progression : a -> b -> c -> d. In this case, we are asking dpath to return to us (and dpath has every reason to expect that there should be) the value at a -> b -> c -> d.

However the problem comes in when the searching method gets to the 2nd level of the dictionary. It expects to find a key "b". But it doesn't find "b". This in itself is enough to cause the search to fail. But  the search will fail silently; you'll get nothing back, and you may not be sure why. This is why the InvalidKeyName error is raised; when the searcher finds a key with a path separator in it, it realizes that you might be relying on "b/c" to be a valid token, when it is impossible for "b/c" to be a valid token as long as your separator is "/". Then the path parser realizes that it cannot trust the fact that "/" is a safe token separator, and that the tokenization itself is completely suspect. With no way to confirm whether you actually meant "a" -> "b/c" -> "d", or if you meant "a/b/c" -> "d" or "a/b" -> "c/d", the method choose instead to raise an InvalidKeyName to alert you to the fact that what you are doing is dangerous and likely to yield the wrong result.

I don't have a particularly elegant answer to your problem, except that if you can avoid it at all, you should not use path separators inside the text of your keys. Perhaps you could run a transform over the keys in your dictionary to replace path separator characters with something else, or to create new nested hashes wherever the path separator is present. But I don't think the solution to this is to make dpath accept separators in keys.

@calebcase , any thoughts on this?

**Comment by akesterson** [comment]
Actually, we could just make the util functions accept paths as an array, rather than a string, which would make the separator meaningless in this context, because no tokenization would be required. So instead of:

    dpath.util.get(x, "/a/b/c/d")

... you would get:

    dpath.util.get(x, ['a', 'b/c', 'd'])

... In this case the separator would be ignored during path validation, and InvalidKeyName would no longer be thrown. And this wouldn't be too terribly much work, since paths already get converted into arrays when they are sent through the tokenizer in dpath.path anyways.

**Comment by calebcase** [comment]
+1 That sounds like a nice clean way to add support for this.

**Comment by msabramo** [comment]
:+1: - sounds like a great idea!

**Comment by sochotnicky** [comment]
This indeed sounds like a nice solution which can be done even including backward compat.

**Comment by akesterson** [comment]
Sorry for my laziness. I'll test, implement & publish this feature sometime within the next 24 hours.

**Comment by sochotnicky** [comment]
@akesterson seriously? Don't sweat it. if it was that critical we should have provided at least a pull request with proposed changes. Do what you need to do.

**Comment by akesterson** [comment]
Closing, this has actually been resolved since 02/04/2015.

**Comment by akesterson** [comment]
This was released in 1.3.3 for those still wondering
